Implementing the alias command

theAstralProgrammer0 opened the pull request:
* I added the first three auxiliary functions required to implement an aliasExecutor function that is yet to be programmed.

* I also added a function that finds the length of a string array - a basic word counter - which will be of utility when building the aliasExecutor function.

theAstralProgrammer0 committed:
Added first 3 functions to handle aliases and a string array word counter

nanafox commented on alias.c:
For reference, here's the definition for the `alias_node` struct
```c
/**
 * struct alias_node - a blueprint for alias nodes
 * @name: the name of the alias
 * @value: the command(s) being aliased
 * @next: pointer to the next alias node
 */
typedef struct alias_node
{
	char *name;
	char *value;
	struct alias_node *next;
} alias_node;
```

I see you are using `key` and `value`. Also, what happens when the value of the alias you just found is an alias?
For example:
```bash
ls='/bin/ls'
l='ls'
my_ls='l'
```

Suppose the user inputs `my_ls`, the result of your current logic will be `l` which is not a valid command. You should work on accounting for "chained" aliases. You could look into how recursion might be of help.

theAstralProgrammer0 replied:
* Right. I will refactor the `key` variable to `name`.
* As for _chained aliases_, I would look into it and refactor again.
* Thanks for pointing these out.

nanafox commented on str_arr_size.c:
I'm yet to fully understand what this function does. Could you throw more light on this?

theAstralProgrammer0 replied:
* Right, so the `str_arr_size` function can be used to get the length of a `NULL` terminated _string array_.
* After tokenizing a string into a string array, we can utilize this function to find how many strings exist in the array.
* For instance, in the `aliasExecutor` function I'm currently working on, I can use this function to find the length of any string array returned by the `strtow` function.
* I will be using the `strtow` function heavily, and it would be useful to know the length of the returned string array - particularly when writing `if` conditions

nanafox replied:
Another team will be handling the tokenizer function #9 so you could rest on this one unless you have a specific use case.
You can check out their tokenizer function to learn more about how it could fit into your needs.

theAstralProgrammer0 replied:
* Alright boss, thanks. I'll look into their function. However, their `tokenize` function can only help create the string array. It does not provide the `size` of the array for future use.
* The `str_arr_size` function will help count the number of strings present in a string array that's returned from the  `tokenize` function, which I will need in my function implementation.

theAstralProgrammer0 committed:
Added function prototype: void aliasExecutor(alias_t *, char*[]);
Refactored 'key' to 'name'
Corrected some syntax errors

theAstralProgrammer0 committed:
Made alias.c betty compliant

theAstralProgrammer0 committed:
Added comment in alias.c to highlight how char *arguments[] should be passed from the shell main file

nanafox commented on alias.c:
You might want to check your line 16 well...

Also, I'd like to believe memory is correctly handled in your code logic.

nanafox replied:
Also, in your `addAlias()` function, you keep adding to the beginning of the list which will influence how the list of aliases is printed out when the command `alias` is issued without arguments. Utilize the `tail` pointer, it also gives you an O(1) runtime.

Currently, your code logic does not account for when `alias` is issued without arguments. Look into that, it's supposed to print the list of aliases in that case.

nanafox replied:
Compare how `sh` prints the aliases, it doesn't include the `alias` keyword in the output. Handle the errors properly with the right error codes and messages. Consult with `sh`.

nanafox commented on liteshell.h:
How do we know an error occurred per the prototype of the `aliasExecutor()` function? Particularly when you just need to print the given alias?

Also, how about changing `*[]` to `**`? 😎

theAstralProgrammer0 committed:
Refactored createNode to initAlias
Debugged addAlias function taking into account the tail member of an aliasList
Adjusted aliasExecutor function to account for changes in initAlias, and addAlias functions
Corrected other errors in alias.c

nanafox commented on alias.c:
I have concerns with the alias functions...

Could you share a screenshot of the result after testing this function?
Also, share the file that you used for the testing.

theAstralProgrammer0 replied:
Alright boss.

nanafox commented on liteshell.h:
Looks okay now.

theAstralProgrammer0 committed:
This file contains 4 alias and 1 unalias functions
addAlias
git commit -m

theAstralProgrammer0 committed:
Updated liteshell.h

theAstralProgrammer0 committed:
This file contains all functions used to handle the alias command arguments
tokenizeAliasArguments
handleAliasCommands
handleAlias
main

theAstralProgrammer0 committed:
This file contains all utility functions that support the alias functionality
printAliasList
freeAlias
freeAliasList
freeTokens
